esp/services/ws_workunits/ws_workunitsHelpers.cpp

```diff
@@ -2089,53 +2031,71 @@ void WsWuInfo::getWorkunitThorSlaveLog(const char *groupName, const char *ipAddr
+        readWorkunitLog(ios.get(), buf, outFile);
+    }
+    else
+    {//legacy wuid
+        appendIOStreamContent(buf, ios.get(), forDownload);
+    }
+}
+
+void WsWuInfo::readWorkunitLog(IBufferedFileIOStream* ios, MemoryBuffer& buf, const char* outFile)
+{
+    VStringBuffer startwuid("Started wuid=%s", wuid.str());
+    VStringBuffer endwuid("Finished wuid=%s", wuid.str());
 
-        VStringBuffer startwuid("Started wuid=%s", wuid.str());
-        VStringBuffer endwuid("Finished wuid=%s", wuid.str());
+    bool eof = false;
+    bool outputThisLine = false;
+    unsigned processID = 0;
 
-        const char *sw = startwuid.str();
-        const char *ew = endwuid.str();
+    StringBuffer line;
+    CWsWuFileHelper helper(nullptr);
+    Owned<IFileIOStream> outIOS = helper.createIOStreamWithFileName(outFile, IFOcreate);
+    while (!eof)
+    {
+        line.clear();
+        eof = ios->readLine(true, line);
+        if (line.isEmpty())
+            continue;
 
-        Owned<IFileIOStream> outIOS;
-        if (!isEmptyString(outFile))
+        if (strstr(line, startwuid))
         {
-            CWsWuFileHelper helper(nullptr);
-            outIOS.setown(helper.createIOStreamWithFileName(outFile, IFOcreate));
+            outputThisLine = true;
+            outputALine(line.length(), line.str(), buf, outIOS);
+            processID = readLogProcessID(line);
         }
-        while (!eof)
+        else if (outputThisLine)
         {
-            line.clear();
-            for (;;)
-            {
-                char c;
-                size32_t numRead = ios->read(1, &c);
-                if (!numRead)
-                {
-                    eof = true;
-                    break;
-                }
-                line.append(c);
-                if (c=='\n')
-                    break;
-            }
-            if (strstr(line.str(), sw))
-                include = true;
-            if (include)
-                outputALine(line.length(), line.str(), buf, outIOS);
-            if (strstr(line.str(), ew))
-                include = false;
+            //If the slave is restarted before WU is finished, we cannot find out the "Finished wuid=...".
+            //So, we should check whether the slave is restarting or not.
+            unsigned pID = readLogProcessID(line);
+            if ((pID > 0) && (pID != processID))
+                break;
+
+            outputALine(line.length(), line.str(), buf, outIOS);
         }
+
+        if (strstr(line, endwuid))
+            break;
     }
-    else
-    {//legacy wuid
-        appendIOStreamContent(buf, ios.get(), forDownload);
-    }
+}
+
+unsigned WsWuInfo::readLogProcessID(const char* line)
+{
+    StringArray items;
+    items.appendList(line, " ", true, 4);
+    if (items.ordinality() < 4)
+        return 0; //no processID in this line
+
+    char* eptr = nullptr;
+    const char* ids = items.item(3);
+    unsigned processID = (unsigned)strtol(ids, &eptr, 10);
```

Comment:
Use of atoi() is more standard.

Reply:
Also, this is also quite expensive. ( Copying everyline into N strings, and putting them in an array ).

It would be better to scan to field, and to end of field, then use atoi_l()

Reply:
I am using strtol because it can check any character inside the field just in case.